File: BowlingTest/Game.cs
using System.Collections.Generic;
using System.Linq;

namespace BowlingTest
{
    public class Game
    {
        private int _finalScore = 0;

        public void roll(int i)
        {
            _frames.Last().roll(i);
            if (_frames.Last().Rolls.Count == 2)
            {
                _frames.Add(new Frame());
            }

            var score = CalcScore(0, _frames);
            _finalScore = score;
        }
        
        private int CalcScore(int lastScore, IEnumerable<Frame> frames)
        {
            if (frames.Any())
            {
                var newScore =  _rules.First(r => r.match(frames))
                    .compute(frames, lastScore);
            
                return CalcScore(newScore, frames.Skip(1));
            }
            return  lastScore;
        }
       
    
        private IRule[] _rules;

        private List<Frame> _frames;

        public Game()
        {   
            _rules = new IRule[]
            {
                new SpareRule(), 
                new BasicRoll()
            };
            _frames = new List<Frame>()
            {
                new Frame()
            };

        }

        public int score()
        {
            return _finalScore;
        }
    }
}

File: src_python/game.py
from basicRoll import BasicRoll

class Frame(object):

    def __init__(self) :
        self.__rolls = []

    def roll(self, i):
        self.__rolls.append(i)

    def get_rolls(self): 
        return self.__rolls[:]

    def count(self) :
        return len(self.__rolls)

class Game(object) : 

    def __init__(self):

        self.__finalScore = 0
        self.__rules = [

            BasicRoll()
        ]

        self.__frames = [Frame()]
        


    def roll(self, i):
        self.__frames[-1].roll(i)

        if self.__frames[-1].count() == 2 :
        
            self.__frames.append(Frame())
        

        score = 0
        for frame in self.__frames :
        
            score = [r for r in self.__rules if r.match(frame)][0].compute(frame, self.followingFrames(frame), score)
        
        self.__finalScore = score

    def score(self):
        return self.__finalScore

    def followingFrames(self, frame):

        return []